tools/topology2/sof-cnl-nocodec.conf

```diff
@@ -0,0 +1,94 @@
+
+#
+# Simple Machine - High level topology - Maps to machine driver.
+#
+#
+# PCM 0 <-> bufferN.0 <-> volumeN.0 <-> bufferN.1 -> SSP0
+#
+
+<include/common/tokens.conf>
+<include/platform/intel/bxt.conf>
+<include/pipelines/pipeline-volume-playback.conf>
+<include/pipelines/pipeline-volume-capture.conf>
+<include/common/connection.conf>
+<include/common/pipe_endpoint.conf>
+<include/components/dai.conf>
+<include/common/data.conf>
+<include/dais/ssp.conf>
+<include/common/manifest.conf>
+
+#
+# Pipeline definitions
+#
+
+# Pipeline ID:1 PCM ID: 0
+Object.volume-playback."1" {
+	pipeline_id	1
+	pcm_id		0
+	pcm_name	"Port0"
+	format		"s24le"
+
+	# set the mixer name for PGA.0
+	pga.0.mixer.0.name "1 Master Playback Volume"
+}
+
+# Pipeline ID:2 PCM ID: 0
+Object.volume-capture."2" {
+	pipeline_id	2
+	pcm_id		0
+	pcm_name	"Port0"
+	format		"s24le"
+
+	# set the mixer name for PGA.0
+	pga.0.mixer.0.name	"2 Master Capture Volume"
+}
+
+#
+# List of all DAIs
+#
+#SSP Index: 0, Direction: duplex
+Object.SSP."0" {
+	direction		"duplex"
+	stream_name		"NoCodec-0"
+	id 			0
+	default_hw_config	0
+	format			"s24le"
+	quirks			"lbm_mode"
+	sample_bits		24
+	Object.hw_config."0" {
+		id		0
+		mclk_freq	24000000
+		bclk_freq	4800000
+		tdm_slot_width	25
+	}
+
+	#Add DAI widgets
+	Object.dai."playback" {
+		direction	"playback"
+		widget_type	"dai_in"
+		period_source_count	2
+		period_sink_count	0
+	}
+
+	Object.dai."capture" {
+		direction	"capture"
+		widget_type	"dai_out"
+		period_source_count	0
+		period_sink_count	2
+	}
```

Comment:
Could you explain what exactly "embedding" those 3 objects here mean?

Reply:
sure, a DAI needs a hw_config and 1/2 DAI widgets depeding on whether it is uni-directional or duplex. And these 3 opbjects are for those.